docs/index.bs

```diff
@@ -2982,35 +2991,41 @@ spec: webappsec-referrer-policy; urlPrefix: https://w3c.github.io/webappsec-refe
           1. Return null.
       1. If |activeWorker|'s <a>state</a> is "`activating`", wait for |activeWorker|'s <a>state</a> to become "`activated`".
       1. If the result of running the [=Run Service Worker=] algorithm with |activeWorker| is *failure*, then set |handleFetchFailed| to true.
-      1. Else [=queue a task=] |task| to run the following substeps:
-          1. Let |e| be the result of <a>creating an event</a> with {{FetchEvent}}.
-          1. Let |requestObject| be a new {{Request}} object associated with |request| and a new associated {{Headers}} object whose [=guard=] is "`immutable`".
-          1. Initialize |e|’s {{Event/type}} attribute to {{fetch!!event}}.
-          1. Initialize |e|’s {{Event/cancelable}} attribute to true.
-          1. Initialize |e|’s {{FetchEvent/request}} attribute to |requestObject|.
-          1. Initialize |e|’s {{FetchEvent/preloadResponse}} to |preloadResponse|.
-          1. Initialize |e|'s {{FetchEvent/clientId}} attribute to |client|'s [=environment/id=].
-          1. If |request| is a <a>non-subresource request</a> and |request|'s [=request/destination=] is not {{RequestDestination/"report"}}, initialize |e|'s {{FetchEvent/resultingClientId}} attribute to |reservedClient|'s [=environment/id=], and to the empty string otherwise.
-          1. If |request| is a <a>navigation request</a>, initialize |e|'s {{FetchEvent/replacesClientId}} attribute to |request|'s [=request/replaces client id=], and to the empty string otherwise.
-          1. <a>Dispatch</a> |e| at |activeWorker|'s [=service worker/global object=].
-          1. Invoke [=Update Service Worker Extended Events Set=] with |activeWorker| and |e|.
-          1. If |e|'s [=FetchEvent/respond-with entered flag=] is set, set |respondWithEntered| to true.
-          1. If |e|'s [=FetchEvent/wait to respond flag=] is set, then:
-              1. Wait until |e|'s [=FetchEvent/wait to respond flag=] is unset.
-              1. If |e|'s [=FetchEvent/respond-with error flag=] is set, set |handleFetchFailed| to true.
-              1. Else, set |response| to |e|'s [=FetchEvent/potential response=].
-          1. If |e|'s <a>canceled flag</a> is set, set |eventCanceled| to true.
-          1. If |fetchInstance| is [=fetch/terminated=], then [=queue a task=] to [=AbortSignal/signal abort=] on |requestObject|'s {{Request/signal}}.
-
-          If |task| is discarded, set |handleFetchFailed| to true.
-
-          The |task| *must* use |activeWorker|'s <a>event loop</a> and the <a>handle fetch task source</a>.
+      1. Else:
+          1. Let |workerRealm| be the [=relevant realm=] of the |activeWorker|'s [=service worker/global object=].
+          1. Set |eventHandled| to [=a new promise=] in |workerRealm|. In following steps, [=resolve=] or [=reject=] |eventHandled| in |workerRealm| only. If |eventHandled| is null, ignore the [=resolve=] or [=reject=] operations on it.
```

Comment:
> If |eventHandled| is null, ignore the [=resolve=] or [=reject=] operations on it.

I'm not sure we should hand-wave like this. Can we just perform a null check when we try to do stuff with this object?

> In following steps, [=resolve=] or [=reject=] |eventHandled| in |workerRealm| only.

I don't think this really works, since it doesn't matter which realm resolves/rejects the promise. The realm only matters for objects.

In this case, you'd just change the instances of

"[=reject=] |eventHandled| with a "{{NetworkError}}" {{DOMException}}"

to

"[=reject=] |eventHandled| with a "{{NetworkError}}" {{DOMException}} **in |workerRealm|**"

It's the error object that gets the realm.